File: Leetcode/79_word_search.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


// Keynote: 
//     當找到第一個letter之後就開始對接下來的字做DFS, helper的d用來track比較到第幾個letter, x,y則是track boad裡面的index,
//     當out of range 就不用繼續找, 當letter跟下一個不一樣也不用繼續找了,找到那自之後先把board[x][y]設成'0',以免重複找,當找完再把這letter回復


class Solution
{
public:
    bool exist(vector<vector<char>> &board, string word)
    {
        if(board.size() == 0) return false;
        bool ans = false;
        row = board.size();
        col = board[0].size(); 
        for(int i = 0; i < row; i ++)
        {
            for(int j = 0; j < col; j++)
            {
                if (helper(board, word, 0, i, j)) return true;
            }
        }

        return false;
    }

    // word記得用const string &, 沒有的話速度差很多
    bool helper(vector<vector<char>> &board, const string& word, int d, int x, int y)
    {   
        if( x < 0 || x >= row || y < 0 || y >= col || board[x][y] != word[d]) return false;
        if (d == word.length() - 1) return true;
        char ch = board[x][y];
        board[x][y] = '0';
        bool result = helper(board, word, d + 1, x - 1, y) ||
            helper(board, word, d + 1, x + 1, y) ||
            helper(board, word, d + 1, x, y - 1) ||
            helper(board, word, d + 1, x, y + 1);
        board[x][y] = ch;

        return result;
    }
private: 
    int row;
    int col;
};


int main()
{
    vector<vector<char>> test{
        { 'A', 'B', 'C', 'E' }, 
        { 'S', 'F', 'C', 'S' }, 
        { 'A', 'D', 'E', 'E' }
    };

    string test_str = "ABCCED";
    Solution sol;
    cout << sol.exist(test, test_str) << endl;
}

File: Leetcode/79_word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])



        def dfs(i: int, j: int, k: int):
            # board[i][j] != word[k]的check 要在最後一步, 不然可能會報index error
            if (i < 0 or i >= m or j <0 or j >=n or (board[i][j] != word[k]) ):
                return False
            
            # 這一步不能放在function 最前面, 不然還沒檢查最後一個單字就return True了
            if k == (len(word) - 1):
                return True 

            # 這一步要在邊界檢查之後, 不然如果i或j是-1, 就會抓取最後一行或最後一列, 而不會報錯
            tmp = board[i][j]
            
            
            board[i][j] = '#'

            found = ( dfs(i-1, j, k+1) or dfs(i, j-1, k+1) or dfs(i+1, j, k+1) or dfs(i, j+1,k+1))

            board[i][j] = tmp
            
            return found
        


        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True 
        return False


優化:
1. 字元頻率與長度預檢查（Character Frequency Check）：
在進入 DFS 前，先統計 board 與 word 的字元出現次數。若字串長度超過棋盤總格數，
或棋盤內某字元的數量少於 word 所需數量，直接回傳 False，完全省去遞迴開銷。

2. 稀有字元起點優化（Reverse Word Search）：
比較 word[0] 與 word[-1] 在棋盤中出現的次數。若開頭字母在棋盤中出現的次數比結尾字母多，
直接將字串反轉（word = word[::-1]）再搜尋。從出現頻率較低的字母當作 DFS 起點，能大幅降低搜尋樹的分支數量。

code: 將以下加入dfs 前
# 1. 長度剪枝
        if len(word) > m * n:
            return False

        # 2. 詞頻剪枝
        board_counts = Counter(char for row in board for char in row)
        word_counts = Counter(word)
        for char, count in word_counts.items():
            if board_counts[char] < count:
                return False

        # 3. 起點稀有度優化（若結尾字元較少，反轉搜尋）
        if board_counts[word[0]] > board_counts[word[-1]]:
            word = word[::-1]